### DISCUSSION TITLE
Is there a way to pull in an elrepo kernel-ml during upgrade?

### DISCUSSION THREAD

**Discussion by rrauenza** [oamg/leapp-repository#1271] [Q&A] [answered]
I've gotten my CentOS 7 painted into a corner where I really need a kernel-ml during and after the upgrade  ....

Is there a simple way to patch in switching from the distro's default kernel to one from elrepo?  I tried just swapping out the vmlinuz-upgrade in /boot, but for some reason the upgrade process also needed iscsi ...

**Comment by pirat89** [comment, marked as answer]
Hi @rrauenza, I am not familiar with this kernel. Well, in case you want to use this specific kernel instead of the standard one, you will need to update possibly more things. note that iscsi is not supported for upgrades right now if you need it.

In general, I suggest to look at actors here:
* https://github.com/oamg/leapp-repository/tree/master/repos/system_upgrade/common/actors/initramfs

When you want to use a specific kernel for the upgrade process, you need to build also the upgrade initramfs right for this kernel so both are compatible (I expect that kernel-ml is not of the same version-release as the RH official one). In short:
* you need to install the kernel into target userspace container first
* build the upgrade initramfs for the specified kernel
* copy both to /boot
* and also you will need to ensure that the upgraded system is going to boot to kernel-ml

The first step can be done in several ways, however, as you will have to modify the code anyway, I suggest you to look at this one actor and modify it properly:
* https://github.com/oamg/leapp-repository/tree/master/repos/system_upgrade/common/actors/initramfs/upgradeinitramfsgenerator

You are interested about `prepare_userspace_for_initram` - see the packages variable. Add the expected names of kernel-ml rpms into the set before the call of `_install_initram_deps`.

Then update `_get_target_kernel_version` to look for the specific kernel rpm. This would be possibly all what is needed regarding the upgrade initramfs + kernel. If not, most likely you will need to update the `generate-initram.sh` script too.

Similarly, you need to ensure that correspongind rpms are installed also during the upgrade. Regarding that part, search for all related to `InstalledTargetKernelInfo` msg. you should find all the code which you will need possibly to update too. There is already solution for that how we handle upgrades with `kernel-rt` (or `kernel-core-rt`). This could guide you how you could modify the solution to deal with kernel-ml.

**Comment by rrauenza** [comment]
Thank you. I'm poking at this today.  CentOS7 started with supporting btrfs, there were some bugs, and I started moving to the kernel-ml line of kernels from elrepo to work around them.  

I'll record my progress here for anyone else needing to do this.

**Reply by pirat89** [reply]
thanks!
